python/BioSimSpace/Sandpit/Exscientia/Parameters/_parameters.py

```diff
@@ -1111,3 +752,195 @@ def _validate(molecule=None, tolerance=None, max_distance=None,
+               own naming scheme, e.g. { "charge" : "my-charge" }
+
+           Returns
+           -------
+
+           molecule : :class:`Molecule <BioSimSpace._SireWrappers.Molecule>`
+               The parameterised molecule.
+        """
+        return _parameterise_openff(molecule, name, work_dir, property_map)
+    return _function
+
+# Dynamically create functions for all of the supported AMBER protein force fields.
+for _ff in _amber_protein_forcefields.keys():
+    # Generate the function and bind it to the namespace.
+    _function = _make_amber_protein_function(_ff)
+    setattr(_namespace, _ff, _function)
+
+    # Expose the function to the user.
+    __all__.append(_ff)
+
+# Expose the two GAFF functions.
+__all__.append("gaff")
+__all__.append("gaff2")
+
+# Create a list of the force field names (to date.)
+# This needs to come after all of the force field functions.
+_forcefields = []           # List of force fields (actual names).
+_forcefields_lower = []     # List of lower case names.
+_forcefield_dict = {}       # Mapping between lower case names and functions.
+for _ff in list(_amber_protein_forcefields.keys()) + ["gaff", "gaff2"]:
+    _forcefields.append(_ff)
+    _forcefields_lower.append(_ff.lower())
+    _forcefield_dict[_ff.lower()] = getattr(_namespace, _ff)
+
+# Dynamically create functions for all available force fields from the Open
+# Force Field Initiative.
+_openforcefields = _try_import("openforcefields")
+
+if _have_imported(_openforcefields):
+    from glob import glob as _glob
+    import os as _os
+    _openff_dirs = _openforcefields.get_forcefield_dirs_paths()
+    _open_forcefields = []
+    # Loop over all force field directories.
+    for _dir in _openff_dirs:
+        # Glob all offxml files in the directory.
+        _ff_list = _glob(f"{_dir}" + "/*.offxml")
+        for _ff in _ff_list:
+            # Get the force field name (base name minus extension).
+            _base = _os.path.basename(_ff)
+            _ff = _os.path.splitext(_base)[0]
+
+            # Only include unconstrained force-fields since we need to go via
+            # an intermediate ParmEd conversion. This means ParmEd must receive
+            # bond parameters. We can then choose to constrain later, if required.
+            # See, e.g: https://github.com/openforcefield/openff-toolkit/issues/603
+
+            if "unconstrained" in _ff:
+                # Append to the list of available force fields.
+                _forcefields.append(_ff)
+                _open_forcefields.append(_ff)
+
+                # Create a sane function name, i.e. replace "-" and "."
+                # characters with "_".
+                _func_name = _ff.replace("-", "_")
+                _func_name = _func_name.replace(".", "_")
+
+                # Generate the function and bind it to the namespace.
+                _function = _make_openff_function(_ff)
+                setattr(_namespace, _func_name, _function)
```

Comment:
I think this dynamic creation has improved on the previous version of the code by a wide margin. I guess one downside of using `setattr` is that it may not play very well with IDE/Jupyter linting so these functions might not be suggested which could be confusing for the user. Not sure if we should run `_function = _make_openff_function(_ff)` explicitly for each force field to enable proper linting - it's up to you.

Reply:
It works fine in Jupyter. I've not tested in an IDE, but take your point. Will see what happens.